src/phl/ecPhlCodeTx.ml

```diff
@@ -136,68 +137,140 @@ let t_set_r side cpos (fresh, id) e tc =
+        let asgn =
+          match lv with
+          | LvVar (x, ty) -> 
+            (* check if first iteration or if variable is being propagated *)
+            let () = assert (asg = PV.empty || PV.mem_pv env x asg) in
+            [(x, ty, e)]
+          | LvTuple xs -> begin
+              match e.e_node with
+              | Etuple es -> 
+                (* check if first iteration or if all the variables in the tuple are being propagated *)
+                let () = assert (asg = PV.empty || 
+                    List.for_all (fun x -> PV.mem_pv env x asg) (List.fst xs)) in
+                List.map2 (fun (x, ty) e -> (x, ty, e)) xs es
+              | _ -> assert false
+          end
+        in
+          (asgn, tl)
+      end
+
+      | _ -> 
+          tc_error pf "cannot find a left-value assignment at given position"
+    in
 
-    | _ ->
-        tc_error pf "cannot find a left-value assignment at given position"
-  in
+    (* Check that we are assigning to local variables only *)
+    List.iter
+      (fun (x, _, _) ->
+        if is_glob x then
+          tc_error pf "left-values must be local variables")
+      asgn;
+
+    (* Check that we are assigning an expression that we can evaluate to a constant  *)
+    let expr_is_closed (e: expr) : bool = e_fv e = Mid.empty && e_read env e = PV.empty in
+    List.iter
+      (fun (_, _, e) ->
+          if not (expr_is_closed e) then
+            tc_error pf "right-values are not closed expression")
+      asgn;
+
+    (* List of variables we are assigning to *)
+    let asg = List.fold_left
+                (fun pv (x, ty, _) -> EcPV.PV.add env x ty pv)
+                asg asgn
+    in
 
-  let (tl1, tl2) =
-    match olen with
-    | None      -> (tl, [])
-    | Some olen ->
-        if List.length tl < olen then
-          tc_error pf "expecting at least %d instructions after assignment" olen;
-        List.takedrop olen tl
-  in
+  
+    let subst =
+      List.fold_left
+        (fun subst (x, _ty, e) ->  Mpv.add env x e subst)
+        subst asgn
+    in
 
-  List.iter
-    (fun (x, _, _) ->
-      if is_glob x then
-        tc_error pf "left-values must be local variables")
-    asgn;
-
-  List.iter
-    (fun (_, _, e) ->
-        if e_fv e <> Mid.empty || e_read env e <> PV.empty then
-          tc_error pf "right-values are not closed expression")
-    asgn;
-
-  let wrs = is_write env tl1 in
-  let asg = List.fold_left
-              (fun pv (x, ty, _) -> EcPV.PV.add env x ty pv)
-              EcPV.PV.empty asgn
-  in
+    
+    let do_subst = Mpv.isubst env subst in
+    let do_substs = Mpv.issubst env subst in
+
+    (* Rebuild assignments *)
+    let asgn_instrs_from_pv_and_subst (asg) (subst) : instr = 
+      let asgns = fst (PV.elements asg) in
+      match asgns with 
+      | [] -> assert false (* meaning we dont propagate, should never happen *)
+      | (pv, ty)::[] -> i_asgn (LvVar (pv, ty), e_var pv ty) |> (Mpv.isubst env subst)
```

Comment:
Use `[x]`, not `x :: []`.

Coding style: we tend to put spaces around binary operators.